tests.texlua/memoize-extract.lua

```diff
@@ -0,0 +1,1279 @@
+		if self.file then
+			short = short:gsub("\\", "\\string\\")
+			long  = long:gsub("\\", "\\string\\")
+			self.file:write(ERROR[format]{short=short, long=long, package_name=package_name})
+		end
+		-- set the exitcode this way
+		exit.succ = exit.error
+	end
+	logging.error = logging._error
+
+	---Log a warning
+	---@param text string
+	---@param quiet boolean
+	---@param format string
+	function logging:_warn(text, quiet, format)
+		format = format or "None"
+		if not quiet then
+			print(WARNING.None{text=text, header=self.header, indent=self.indent})
+		end
+		if self.file then
+			text = text:gsub("\\", "\\")
+			self.file:write(WARNING[format]{text=text, texindent=self.texindent, package_name=self.package_name})
+		end
+		-- set the exitcode this way
+		exit.succ = exit.warn
+	end
+	logging.warn = logging._warn
+
+	---Log info message
+	---@param text string
+	---@param quiet boolean
+	---@param format string
+	function logging:_info(text, quiet, format)
+		format = format or "None"
+		if not quiet then
+			print(INFO.None{text=text, header=self.header, indent=self.indent})
+		end
+		if self.file then
+			text = text:gsub("\\", "\\")
+			self.file:write(INFO[format]{text=text, texindent=self.texindent, package_name=self.package_name})
+		end
+	end
+	logging.info = logging._info
+end
+
+-- "forward declarations" for logging versions of error/assert
+local log_assert
+local log_error
+
+---analog to lua's assert, define a function which uses logging for the message instead
+---@param cond boolean condition to be checked by this assertion
+---@param msg string? message shown when the assertion fails
+---@param cleanup fun()? function invoked after logging the message used for additional cleanup (can still use logging, the log-file is not yet closed). Might be omitted
+log_assert = function(cond, msg, cleanup)
+	if not cond then
+		logging:error("", msg or "")
+		if cleanup then cleanup() end
+		logging:close()
+		exit.error()
+	end
+end
+
+---analog to lua's error, define a function which uses logging for the message instead
+---@param msg string? message shown
+---@param cleanup fun()? function invoked after logging the message used for additional cleanup (can still use logging, the log-file is not yet closed). Might be omitted
+log_error = function(msg, cleanup)
+	logging:error("", msg)
+	if cleanup then cleanup() end
+	logging:close()
+	exit.error()
+end
+
+---Unquote a quoted string
+---@param fn string quoted filename
+---@return string
+local function unquote(fn)
+	local r = fn:gsub("\"(.-)\"", "%1")
+	return r
+end
+
+local md5pat = ("%x"):rep(32)
+--- Parses the extern_path
+-- in python this is a simple regex, but lua patterns cannot do the same things,
+-- so we need multiple ones
+---@param path string
+---@return string? dir_prefix
+---@return string? name_prefix
+---@return string? code_md5sum
+---@return string? context_md5sum
+local function parse_extern_path(path)
+	-- TODO maybe lpeg would be better suited for parsing this
```

Comment:
LPEG would indeed be better here, but what you have right now seems fine to me too.